src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs

```diff
@@ -167,7 +170,25 @@ private CacheableKeyRing CreateCacheableKeyRingCoreStep2(DateTimeOffset now, Can
 
         _logger.UsingKeyAsDefaultKey(defaultKey.KeyId);
 
-        var nextAutoRefreshTime = now + GetRefreshPeriodWithJitter(KeyManagementOptions.KeyRingRefreshPeriod);
+        // If we're in a scenario where we think other app instances may be generating keys that will be
+        // activated before they are propagated, we'll use a shorter refresh period to ensure that we pick
+        // up those keys so that we can select from the same pool of candidates as the other instances.
+        // 
+        // These not-yet-propagated keys are usually immediately-activated, but an app restarted after a period
+        // of downtime may discover that it has a valid, but soon-to-be-expired key. The replacement will not
+        // be immediately-activated, but may be activated before it has propagated.
+        var useShortRefreshPeriod = allowShortRefreshPeriod &&
+            (generatedKey is not null
```

Comment:
Nit: Move the second clause of this Boolean to a well-defined method (e.g. `GeneratedKeyWithinPropogationWindow` or something).